### PR TITLE
Add bind9

### PR DESCRIPTION
**Pull request by belka-ew**

### CODE REVIEW

**Comment by mumahendras3** [comment]
Hi, thanks for this submission!

A have briefly read `named` man page and I discover there is `-g` option to redirect all logs to stderr instead of syslog. I would prefer to use that instead of `-f` since with `-g` we can create a logger service named `bind-log` which will log all `named` messages and save them in a separate logging directory complete with automatic logs rotation using `s6-log`.

I have pushed `templates` directory to the master branch where you can find a directory named `logger` in it. Would you mind creating `bind-log` service using that template? The template should be easy to modify, you can just run `sed -i "s/svname/bind/g"` to all files inside `logger` directory that contains "svname" in it (`consumer-for`, `run`, and `env/LOGDIR_PATH`) and then copy that directory to `rc.bind/bind-log`. Lastly, don't forget to add a new file `producer-for` in `rc.bind/bind` with the content `bind-log`.

Again, thanks for your time!

**Comment by belka-ew** [comment]
bind has its own logger which is more customizable than stderr and syslog (similar to apache). It can write different kinds of logs to different files and it supports log rotation.

The configuration looks something like this (/etc/named.conf):

```
logging {
     channel default_log {
          file "/var/log/named/default" versions 3 size 20m;
          print-time yes;
          print-category yes;
          print-severity yes;
          severity info;
     };
     channel auth_servers_log {
          file "/var/log/named/lauth_servers" versions 3 size 20m;
          print-time yes;
          print-category yes;
          print-severity yes;
          severity info;
     };
     category default { default_log; };
     category resolver { auth_servers_log;};
};
```

Here we write different kinds of messages to two files; we keep 3 versions of the logs, 20 MB each.

This is the reason I left out the logger. But if you think it would be better to have `bind-log`, I'll add one.

**Comment by mumahendras3** [comment]
After some little experimenting, I suggest we should still create `bind-log` and then use `-f -L /proc/self/fd/2` for `env/CMD_OPTS` instead. This way, every logs that are not configured in `/etc/named.conf` will be redirected to `bind-log` instead of syslog.

How about that?

File: source/rc.bind/bind/run [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# Start the BIND name server daemon (named) #
+#############################################
+
+fdmove -c 2 1
+s6-envdir env
+multisubstitute {
+  importas -u CMD CMD
+  importas -usCd" " CMD_OPTS CMD_OPTS
+}
+s6-notifyoncheck -dn0 -c "rndc status"
```
Comment (mumahendras3) [inline source/rc.bind/bind/run]: Will there be other service that depends on this service? I would prefer to not use `s6-notifyoncheck` until there are other services that depend on this and need a proper readiness notification from this service. Also `rndc` will exit non-zero and spew errors when neither `/etc/rndc.conf` nor `/etc/rndc.key` was found, spamming `bind-log` with error messages and blocking `s6-rc -u change bind` invocation by the user as if `named` hasn't started yet while in reality it is actually already running. Does `named` still function correctly without `rndc.key` or `rndc.conf`? I tried running `named` without them and it still runs without error, just a notice in the logs like `21-Dec-2020 22:54:34.980 general: notice: couldn't add command channel 127.0.0.1#953: file not found`. Although I never used `named` before so I don't know if it actually functions correctly or not.

File: source/rc.bind/bind-log/env/LOG_SCRIPT [hunk 2]
```diff
@@ -0,0 +1,2 @@
+n4 s500000 T $LOGDIR_PATH
```
Comment (mumahendras3) [inline source/rc.bind/bind-log/env/LOG_SCRIPT]: Since `named` already provides its own timestamp, we don't need the `T` directive here.

**Review by mumahendras3 — CHANGES_REQUESTED** [review]
The `run` script inside `bind` and `bind-log` doesn't need to be marked as executable for s6-rc service definitions so don't forget to `chmod -x` those two's run script.

Also don't forget to add the file `producer-for` in `bind` with the content `bind-log` since `s6-rc-compile` will throw error if that doesn't exist.

Thanks!

File: source/rc.bind/bind/run (hunk 1 above)
Comment (belka-ew) [reply source/rc.bind/bind/run]: Good point. It's been a long time since I've configured bind. If I remember correctly rndc isn't required and named would function correctly without `rndc.key` and `rndc.conf`. `rndc.key`/`rndc.conf` are required to give `rndc` access to `named`, because `rndc` can control `named` over network, they don't always run on the same machine.

> Will there be other service that depends on this service?

Probably not on stock slackware, but I can imagine services that depend on bind. I'm using bind because of rspamd. rspamd does a lot of DNS queries, therefore it requires a caching DNS server like bind. And in my setup bind is a dependency of rspamd. If bind is used for more than caching, then everything that should resolve external domains, would depend on bind.

I'm not sure what is a better way to make a health check. `named` creates a pid file in `/run/named`, but it can be disabled (and I do so, when using s6). `named` creates `/run/named/session.key`, but it can be disabled too, but the file is created with standard configuration. I picked the `rndc` check just because it is the tool for this job. `rndc` is the tool to query and check and control `named`, so it's kind of created for checking `named` status.

Basically I can remove the check. `bind` starts pretty fast and nothing bad happens if it's not yet running, probably some domain names can't be resolved (probably worse for oneshots than longruns). What do you think?

Comment (mumahendras3) [reply source/rc.bind/bind/run]: > Probably not on stock slackware, but I can imagine services that depend on bind. I'm using bind because of rspamd. rspamd does a lot of DNS queries, therefore it requires a caching DNS server like bind. And in my setup bind is a dependency of rspamd. If bind is used for more than caching, then everything that should resolve external domains, would depend on bind.
> 
> I'm not sure what is a better way to make a health check. `named` creates a pid file in `/run/named`, but it can be disabled (and I do so, when using s6). `named` creates `/run/named/session.key`, but it can be disabled too, but the file is created with standard configuration. I picked the `rndc` check just because it is the tool for this job. `rndc` is the tool to query and check and control `named`, so it's kind of created for checking `named` status.
> 
> Basically I can remove the check. `bind` starts pretty fast and nothing bad happens if it's not yet running, probably some domain names can't be resolved (probably worse for oneshots than longruns). What do you think?

Thanks for your explanations, we'll use `s6-notifyoncheck` then. To avoid `s6-rc -u change` from blocking when the user hasn't configured `rndc` yet, we'll do something like below for the run script:

```
#!/bin/execlineb -P

#############################################
# Start the BIND name server daemon (named) #
#############################################

fdmove -c 2 1
s6-envdir env
multisubstitute {
  importas -u CMD CMD
  importas -usCd" " CMD_OPTS CMD_OPTS
}
ifelse { [ -r /etc/rndc.conf -o -r /etc/rndc.key ] }
{
  s6-notifyoncheck -dn0 -c
  "redirfd -w 2 /dev/null
  pipeline { rndc -s localhost status }
  grep -q \"server is up and running\""
  exec -c $CMD $CMD_OPTS
}
# We can't use rndc to poll the named server readiness so we'll announce
# this service's readiness after some fixed period of time
background -d {
  if { sleep 1 }
  fdmove 1 3
  echo
}
unexport ! fdclose 3
exec -c $CMD $CMD_OPTS
```

So instead of directly go to `s6-notifyoncheck`, we'll test first the existence of `/etc/rndc.conf` or `/etc/rndc.key` and then use `rndc` to poll readiness if one of them exists. Here I change the check script to pipe the output of `rndc` to grep and look for the string "server is up and ready". When `rndc` hasn't configured yet, we'll just start `named` immediately and then wait for 1 second. After that 1 second elapsed, print a newline to the notification fd to announce service readiness.

Will that be good enough for you?

Comment (belka-ew) [reply source/rc.bind/bind/run]: Looks good.

Since
```
"redirfd -w 2 /dev/null
  pipeline { rndc -s localhost status }
  grep -q \"server is up and running\""
```
is a long string, can I put it into `data/check`? And if I'm not mistaken s6-rc doesn't mark `data/check` as executable automatically.

Comment (mumahendras3) [reply source/rc.bind/bind/run]: > Since
> 
> ```
> "redirfd -w 2 /dev/null
>   pipeline { rndc -s localhost status }
>   grep -q \"server is up and running\""
> ```
> 
> is a long string, can I put it into `data/check`?

Yes, feel free to do that.

> And if I'm not mistaken s6-rc doesn't mark `data/check` as executable automatically.

Yes, since s6-rc will only copy the `data` directory verbatim.
